File: examples/atomicLayerDeposition/atomicLayerDeposition.py
import viennacs2d as vcs


# Config file reader helper function
def ReadConfigFile(fileName: str):
    """Read a config file in the ViennaPS standard config file format.

    Parameters
    ----------
    fileName: str
              Name of the config file.

    Returns
    -------
    dict
        A dictionary containing the parameters from the config file.
    """
    par_dict = {}

    with open(fileName, "r") as file:
        lines = file.readlines()
        for line in lines:

            line = line[: line.find("#")]  # remove comments

            if len(line) > 0:
                par_name = line[: line.find("=")].strip(" ")
                par_value = line[line.find("=") + 1 :]

                try:
                    val = float(par_value)
                except:
                    val = par_value

                par_dict[par_name] = val

    return par_dict


def makeLShape(params: dict):
    try:
        import viennals2d as vls
    except ImportError:
        print(
            "The ViennaLS Python library is required to run this example. "
            "Please install it and try again."
        )
        return

    gridDelta = params["gridDelta"]
    bounds = [0] * vcs.D * 2
    bounds[0] = -params["verticalWidth"] / 2.0 - params["xPad"]
    bounds[1] = params["verticalWidth"] / 2.0 + params["xPad"]
    if vcs.D == 3:
        bounds[2] = -params["verticalWidth"] / 2.0 - params["xPad"]
        bounds[3] = (
            -params["verticalWidth"] / 2.0 + params["xPad"] + params["horizontalWidth"]
        )
    else:
        bounds[1] = (
            -params["verticalWidth"] / 2.0 + params["xPad"] + params["horizontalWidth"]
        )

    boundaryCons = [vls.BoundaryConditionEnum.REFLECTIVE_BOUNDARY] * (vcs.D - 1)
    boundaryCons.append(vls.BoundaryConditionEnum.INFINITE_BOUNDARY)

    substrate = vls.Domain(bounds, boundaryCons, gridDelta)
    normal = [0.0] * vcs.D
    origin = [0.0] * vcs.D
    normal[vcs.D - 1] = 1.0
    origin[vcs.D - 1] = params["verticalDepth"]
    vls.MakeGeometry(substrate, vls.Plane(origin, normal)).apply()

    # Create the vertical trench
    vertBox = vls.Domain(substrate)
    minPoint = [0] * vcs.D
    maxPoint = [0] * vcs.D
    for i in range(vcs.D - 1):
        minPoint[i] = -params["verticalWidth"] / 2.0
        maxPoint[i] = params["verticalWidth"] / 2.0
    maxPoint[vcs.D - 1] = params["verticalDepth"]
    vls.MakeGeometry(vertBox, vls.Box(minPoint, maxPoint)).apply()
    vls.BooleanOperation(
        substrate, vertBox, vls.BooleanOperationEnum.RELATIVE_COMPLEMENT
    )

    # Create the horizontal trench
    horiBox = vls.Domain(substrate)
    minPoint = [0] * vcs.D
    maxPoint = [params["verticalWidth"] / 2.0] * vcs.D
    for i in range(vcs.D - 1):
        minPoint[i] = -params["verticalWidth"] / 2.0
    maxPoint[vcs.D - 1] = params["horizontalHeight"]
    maxPoint[vcs.D - 2] = -params["verticalWidth"] / 2.0 + params["horizontalWidth"]
    vls.MakeGeometry(horiBox, vls.Box(minPoint, maxPoint)).apply()
    vls.BooleanOperation(
        substrate, horiBox, vls.BooleanOperationEnum.RELATIVE_COMPLEMENT
    )

    return substrate


import viennals2d as vls

# vcs.Logger.setLogLevel(vcs.LogLevel.INTERMEDIATE)
params = ReadConfigFile("config.txt")
# vps.setNumThreads(int(params["numThreads"]))

# Create a cellSet
ls = makeLShape(params)
cellSet = vcs.DenseCellSet()
cellSet.setCellSetPosition(True)
cellSet.setCoverMaterial(1)
matMap = vls.MaterialMap()
matMap.insertNextMaterial(0)
cellSet.fromLevelSets([ls], matMap, params["topSpace"] + params["verticalDepth"])

# Segment the cells into surface, material, and gas cells
vcs.SegmentCells(cellSet, "CellType", 1).apply()

# Calculate the mean free path for the gas cells
mfpCalc = vcs.MeanFreePath(cellSet)
mfpCalc.setNumRaysPerCell(params["raysPerCell"])
mfpCalc.setReflectionLimit(int(params["reflectionLimit"]))
mfpCalc.setRngSeed(int(params["seed"]))
mfpCalc.setMaterial(1)
mfpCalc.setBulkLambda(params["bulkLambda"])
mfpCalc.apply()

cellSet.writeVTU("initial.vtu")

# Run the atomic layer deposition model
model = vcs.AtomicLayerProcess(cellSet)
model.setMaxLambda(params["bulkLambda"])
model.setPrintInterval(params["printInterval"])
model.setStabilityFactor(params["stabilityFactor"])
model.setFirstPrecursor(
    "H2O",
    params["H2O_meanThermalVelocity"],
    params["H2O_adsorptionRate"],
    params["H2O_desorptionRate"],
    params["p1_time"],
    params["inFlux"],
)
model.setSecondPrecursor(
    "TMA",
    params["TMA_meanThermalVelocity"],
    params["TMA_adsorptionRate"],
    params["TMA_desorptionRate"],
    params["p2_time"],
    params["inFlux"],
)
model.setPurgeParameters(params["purge_meanThermalVelocity"], params["purge_time"])
model.setReactionOrder(params["reactionOrder"])
model.apply()

File: examples/atomicLayerDeposition/atomicLayerDeposition.cpp
#include <vcUtil.hpp>

#include <csAtomicLayerProcess.hpp>
#include <csMeanFreePath.hpp>
#include <csSegmentCells.hpp>

#include "geometry.hpp"

namespace cs = viennacs;

int main(int argc, char *argv[]) {
  constexpr int D = 2;
  using NumericType = double;

  // Parse the parameters
  cs::util::Parameters params;
  if (argc > 1) {
    params.readConfigFile(argv[1]);
  } else {
    std::cout << "Usage: " << argv[0] << " <config file>" << std::endl;
    return 1;
  }
  omp_set_num_threads(params.get<int>("numThreads"));

  // Create a cellSet
  auto cellSet = cs::SmartPointer<cs::DenseCellSet<NumericType, D>>::New();
  cellSet->setCellSetPosition(true);

  auto levelSets = makeLShape<NumericType, D>(params);
  // Generate the cell set from the domain
  cellSet->fromLevelSets(levelSets, nullptr,
                         params.get("verticalDepth") + params.get("topSpace"));
  cellSet->writeVTU("initial.vtu");

  // Segment the cells into surface, material, and gas cells
  cs::SegmentCells<NumericType, D> segmentation(cellSet);
  segmentation.setBulkMaterial(1);
  segmentation.apply();

  cs::Timer timer;
  timer.start();

  // Calculate the mean free path for the gas cells
  cs::MeanFreePath<NumericType, D> mfpCalc(cellSet);
  mfpCalc.setNumRaysPerCell(params.get("raysPerCell"));
  mfpCalc.setReflectionLimit(params.get<int>("reflectionLimit"));
  mfpCalc.setRngSeed(params.get<int>("seed"));
  mfpCalc.setMaterial(1);
  mfpCalc.setBulkLambda(params.get("bulkLambda"));
  mfpCalc.apply();

  timer.finish();
  std::cout << "Mean free path calculation took " << timer.totalDuration * 1e-9
            << " seconds." << std::endl;

  cellSet->writeVTU("initial.vtu");

  cs::AtomicLayerProcess<NumericType, D> model(cellSet);
  model.setMaxLambda(params.get("bulkLambda"));
  model.setPrintInterval(params.get("printInterval"));
  model.setStabilityFactor(params.get("stabilityFactor"));
  model.setFirstPrecursor("H2O", params.get("H2O_meanThermalVelocity"),
                          params.get("H2O_adsorptionRate"),
                          params.get("H2O_desorptionRate"),
                          params.get("p1_time"), params.get("inFlux"));
  model.setSecondPrecursor("TMA", params.get("TMA_meanThermalVelocity"),
                           params.get("TMA_adsorptionRate"),
                           params.get("TMA_desorptionRate"),
                           params.get("p2_time"), params.get("inFlux"));
  model.setPurgeParameters(params.get("purge_meanThermalVelocity"),
                           params.get("purge_time"));
  // The deposition probability is (H2O_cov * TMA_cov)^order
  model.setReactionOrder(params.get("reactionOrder"));

  model.apply();

  cellSet->writeVTU("final.vtu");
}
